x-pack/plugin/esql/arrow/src/test/java/org/elasticsearch/xpack/esql/arrow/ArrowResponseTests.java

```diff
@@ -260,45 +279,149 @@ public void testSingleBlock() throws IOException {
         compareEsqlAndArrow(testCase);
     }
 
-    /**
-     * Test that multivalued arrays are rejected
-     */
-    public void testMultivaluedField() throws IOException {
+    public void testMultivaluedPrimitive() throws IOException {
         IntBlock.Builder builder = BLOCK_FACTORY.newIntBlockBuilder(0);
+        builder.beginPositionEntry();
         builder.appendInt(42);
+        builder.appendInt(43);
+        builder.endPositionEntry();
+
+        // The multivalue can be null, but a multivalue cannot contain nulls.
+        // Calling appendNull within a begin/endEntry causes consistency checks to fail in build()
+        // See also https://github.com/elastic/elasticsearch/issues/114324
         builder.appendNull();
+
         builder.beginPositionEntry();
         builder.appendInt(44);
         builder.appendInt(45);
         builder.endPositionEntry();
+
+        // single value
         builder.appendInt(46);
+
         IntBlock block = builder.build();
+        builder.close();
 
         // Consistency check
         assertTrue(block.mayHaveMultivaluedFields());
+        assertEquals(4, block.getPositionCount()); // counts null entries
+        assertEquals(5, block.getTotalValueCount()); // nulls aren't counted
+
+        // Value 0
+        assertEquals(2, block.getValueCount(0));
         assertEquals(0, block.getFirstValueIndex(0));
-        assertEquals(1, block.getValueCount(0));
+        assertEquals(42, block.getInt(block.getFirstValueIndex(0)));
+        assertEquals(43, block.getInt(block.getFirstValueIndex(0) + 1));
 
-        // null values still use one position in the array
+        // Value 1
         assertEquals(0, block.getValueCount(1));
-        assertEquals(1, block.getFirstValueIndex(1));
-        assertTrue(block.isNull(1));
-        assertEquals(0, block.getInt(1));
+        assertTrue(block.isNull(1)); // This is the position index, not value index
+        // No value, but still occupies a value slot with zero
+        assertEquals(2, block.getFirstValueIndex(1));
+        assertEquals(0, block.getInt(block.getFirstValueIndex(1)));
+        assertEquals(3, block.getFirstValueIndex(2));
 
-        assertEquals(2, block.getFirstValueIndex(2));
+        // Value 2
         assertEquals(2, block.getValueCount(2));
-        assertEquals(2, block.getFirstValueIndex(2));
+        assertEquals(3, block.getFirstValueIndex(2));
+        assertEquals(44, block.getInt(block.getFirstValueIndex(2)));
         assertEquals(45, block.getInt(block.getFirstValueIndex(2) + 1));
 
-        assertEquals(4, block.getFirstValueIndex(3));
+        // Value 3
+        assertEquals(1, block.getValueCount(3));
+        assertEquals(5, block.getFirstValueIndex(3));
+        assertEquals(46, block.getInt(block.getFirstValueIndex(3)));
+
+        // End of block
+        assertEquals(6, block.getFirstValueIndex(4));
+
+        var column = TestColumn.create("vector", "integer");
```

Comment:
nit: This sets `column.multivalue` to a random boolean. Shouldn't it be `true` in this test? (I guess this matters only when creating random blocks, but it could still be confusing.)

Reply:
Doh. Good catch! Fixed in 6d79e99292b